File: GLLara/GLLViewDrawer.m
//
//  GLLViewDrawer.m
//  GLLara
//
//  Created by Torsten Kammer on 26.09.12.
//  Copyright (c) 2012 Torsten Kammer. All rights reserved.
//

#import "GLLViewDrawer.h"

#import "NSColor+Color32Bit.h"
#import "GLLAmbientLight.h"
#import "GLLCamera.h"
#import "GLLDirectionalLight.h"
#import "GLLNotifications.h"
#import "GLLView.h"
#import "simd_matrix.h"
#import "simd_project.h"

#import "GLLara-Swift.h"

#import "GLLResourceIDs.h"
#import "GLLRenderParameters.h"
#import "GLLResourceManager.h"

@interface GLLViewDrawer ()
{
	NSArray<NSManagedObject *> *lights; // Always one ambient and three directional ones. Don't watch for mutations.
    
    id<MTLDevice> device;
    id<MTLCommandQueue> commandQueue;
	
    id<MTLBuffer> transformBuffer;
    id<MTLBuffer> lightBuffer;
	BOOL needsUpdateMatrices;
	BOOL needsUpdateLights;
}

- (void)_updateLights;
- (void)_updateMatrices;

@end

@implementation GLLViewDrawer

- (id)initWithManagedSceneDrawer:(GLLSceneDrawer *)drawer camera:(GLLCamera *)camera view:(GLLView *)view;
{
	if (!(self = [super init])) return nil;

    _view = view;
    device = view.device;
    commandQueue = [device newCommandQueue];
    _view.delegate = self;
    
	_camera = camera;
	_sceneDrawer = drawer;
	__weak id weakSelf = self;
	[[NSNotificationCenter defaultCenter] addObserverForName:GLLSceneDrawerNeedsUpdateNotification object:_sceneDrawer queue:[NSOperationQueue mainQueue] usingBlock:^(NSNotification *note){
		[[weakSelf view] setNeedsDisplay:YES];
	}];
	
	lights = [[NSMutableArray alloc] initWithCapacity:4];
	
	// Prepare light buffer.
    lightBuffer = [device newBufferWithLength:sizeof(struct GLLLightsBuffer) options:MTLResourceStorageModeManaged];
    lightBuffer.label = @"global-lights";
	
	// Load existing lights
	NSFetchRequest *allLightsRequest = [[NSFetchRequest alloc] init];
	allLightsRequest.entity = [NSEntityDescription entityForName:@"GLLLight" inManagedObjectContext:drawer.managedObjectContext];
	allLightsRequest.sortDescriptors = @[ [NSSortDescriptor sortDescriptorWithKey:@"index" ascending:YES] ];
	lights = [drawer.managedObjectContext executeFetchRequest:allLightsRequest error:NULL];
	
	NSAssert(lights.count == 4, @"There are not four lights.");
	
	// Register for ambient light color updates
	[lights[0] addObserver:self forKeyPath:@"color" options:0 context:NULL];
	// Register for directional light color updates
	for (int i = 0; i < 3; i++)
		[lights[i + 1] addObserver:self forKeyPath:@"uniformBlock" options:0 context:NULL];
	
	// Transform buffer
    transformBuffer = [device newBufferWithLength:sizeof(mat_float16) options:MTLResourceStorageModeManaged];
    transformBuffer.label = @"global-transform";
	[self.camera addObserver:self forKeyPath:@"viewProjectionMatrix" options:0 context:0];
	
	// Other necessary render state. Thanks to Metal, that got cut down a lot.
    view.clearColor = MTLClearColorMake(0.2, 0.2, 0.2, 1.0);
	
	needsUpdateMatrices = YES;
	needsUpdateLights = YES;
	
	return self;
}

- (void)dealloc
{
	[lights[0] removeObserver:self forKeyPath:@"color"];
	
	for (int i = 0; i < 3; i++)
		[lights[i + 1] removeObserver:self forKeyPath:@"uniformBlock"];
	
	[self.camera removeObserver:self forKeyPath:@"viewProjectionMatrix"];
}

- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context
{
	if ([keyPath isEqual:@"viewProjectionMatrix"])
	{
		needsUpdateMatrices = YES;
		needsUpdateLights = YES;
		self.view.needsDisplay = YES;
	}
	else if ([keyPath isEqual:@"uniformBlock"] || [keyPath isEqual:@"color"])
	{
		needsUpdateLights = YES;
		self.view.needsDisplay = YES;
	}
	else
	{
		[super observeValueForKeyPath:keyPath ofObject:object change:change context:context];
	}
}

- (void)drawShowingSelection:(BOOL)selection into:(id<MTLRenderCommandEncoder>)commandEncoder;
{
	if (needsUpdateMatrices) [self _updateMatrices];
	if (needsUpdateLights) [self _updateLights];
    
    [commandEncoder setVertexBuffer:transformBuffer offset:0 atIndex:GLLVertexInputIndexViewProjection];
    [commandEncoder setVertexBuffer:lightBuffer offset:0 atIndex:GLLVertexInputIndexLights];
    [commandEncoder setFragmentBuffer:lightBuffer offset:0 atIndex:GLLFragmentBufferIndexLights];
	
    [self.sceneDrawer drawWithShowingSelection:selection into:commandEncoder];
}

#pragma mark - Image rendering

- (void)writeImageToURL:(NSURL *)url fileType:(NSString *)type size:(CGSize)size;
{
	NSUInteger dataSize = size.width * size.height * 4;
	void *data = malloc(dataSize);
	[self renderImageOfSize:size toColorBuffer:data];
	
	CFDataRef imageData = CFDataCreateWithBytesNoCopy(kCFAllocatorDefault, data, dataSize, kCFAllocatorMalloc);
	CGDataProviderRef dataProvider = CGDataProviderCreateWithCFData(imageData);
	CFRelease(imageData);
	
	CGColorSpaceRef colorSpace = CGColorSpaceCreateWithName(kCGColorSpaceSRGB);
	CGImageRef image = CGImageCreate(size.width,
									 size.height,
									 8,
									 32,
									 4 * size.width,
									 colorSpace,
									 (CGBitmapInfo) kCGImageAlphaLast,
									 dataProvider,
									 NULL,
									 YES,
									 kCGRenderingIntentDefault);
	
	CGDataProviderRelease(dataProvider);
	CGColorSpaceRelease(colorSpace);
	
	CGImageDestinationRef imageDestination = CGImageDestinationCreateWithURL((__bridge CFURLRef) url, (__bridge CFStringRef) type, 1, NULL);
	CGImageDestinationAddImage(imageDestination, image, NULL);
	CGImageDestinationFinalize(imageDestination);
	
	CGImageRelease(image);
	CFRelease(imageDestination);
}

- (void)renderImageOfSize:(CGSize)size toColorBuffer:(void *)colorData;
{
    /*
	// What is the largest tile that can be rendered?
	[self.context makeCurrentContext];
	GLint maxTextureSize, maxRenderbufferSize;
	glGetIntegerv(GL_MAX_TEXTURE_SIZE, &maxTextureSize);
	glGetIntegerv(GL_MAX_RENDERBUFFER_SIZE, &maxRenderbufferSize);
	// Divide max size by 2; it seems some GPUs run out of steam otherwise.
	GLint maxSize = MIN(maxTextureSize, maxRenderbufferSize) / 4;
	
	// Prepare framebuffer (without texture; a new one is created for every tile)
	GLuint framebuffer;
	glGenFramebuffers(1, &framebuffer);
	glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
	
	GLuint depthRenderbuffer;
	glGenRenderbuffers(1, &depthRenderbuffer);
	glBindRenderbuffer(GL_RENDERBUFFER, depthRenderbuffer);
    
    // Get old viewport. Oh god, a glGet, how slow and annoying
    GLint oldViewport[4];
    glGetIntegerv(GL_VIEWPORT, oldViewport);
	
	// Prepare textures
	GLuint numTextures = ceil(size.width / maxSize) * ceil(size.height / maxSize);
	GLuint *textureNames = calloc(sizeof(GLuint), numTextures);
	glGenTextures(numTextures, textureNames);
	
	// Pepare background thread. This waits until textures are done, then loads them into colorData.
	__block NSUInteger finishedTextures = 0;
	__block dispatch_semaphore_t texturesReady = dispatch_semaphore_create(0);
	__block dispatch_semaphore_t downloadReady = dispatch_semaphore_create(0);
	
	NSOpenGLContext *backgroundLoadingContext = [[NSOpenGLContext alloc] initWithFormat:self.pixelFormat shareContext:self.context];
	
	dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
		[backgroundLoadingContext makeCurrentContext];
		NSUInteger downloadedTextures = 0;
		while (downloadedTextures < numTextures)
		{
			dispatch_semaphore_wait(texturesReady, DISPATCH_TIME_FOREVER);
			
			GLint row = (GLint) downloadedTextures / (GLint) ceil(size.width / maxSize);
			GLint column = (GLint) downloadedTextures % (GLint) ceil(size.width / maxSize);
			
			glPixelStorei(GL_PACK_ROW_LENGTH, size.width);
			glPixelStorei(GL_PACK_SKIP_ROWS, row * maxSize);
			glPixelStorei(GL_PACK_SKIP_PIXELS, column * maxSize);
			
			glBindTexture(GL_TEXTURE_2D, textureNames[downloadedTextures]);
			glGetTexImage(GL_TEXTURE_2D, 0, GL_RGBA, GL_UNSIGNED_BYTE, colorData);
			
			glDeleteTextures(1, &textureNames[downloadedTextures]);
			
			downloadedTextures += 1;
		}
		dispatch_semaphore_signal(downloadReady);
	});
	
	mat_float16 cameraMatrix = [self.camera viewProjectionMatrixForAspectRatio:size.width / size.height];
	
	// Set up state for rendering
	// We invert drawing here so it comes out right in the file. That makes it necessary to turn cull face around.
	glCullFace(GL_FRONT);
	glDisable(GL_MULTISAMPLE);
	
	// Render
	for (NSUInteger y = 0; y < size.height; y += maxSize)
	{
		for (NSUInteger x = 0; x < size.width; x += maxSize)
		{
			// Setup size
			GLuint width = MIN(size.width - x, maxSize);
			GLuint height = MIN(size.height - y, maxSize);
			glViewport(0, 0, width, height);
			
			glBindFramebuffer(GL_FRAMEBUFFER, framebuffer);
			
			// Setup buffers + textures
			glBindTexture(GL_TEXTURE_2D, textureNames[finishedTextures]);
			glTexImage2D(GL_TEXTURE_2D, 0, GL_RGBA8, width, height, 0, GL_RGBA, GL_UNSIGNED_BYTE, NULL);
			glFramebufferTexture2D(GL_FRAMEBUFFER, GL_COLOR_ATTACHMENT0, GL_TEXTURE_2D, textureNames[finishedTextures], 0);
			
			glBindRenderbuffer(GL_RENDERBUFFER, depthRenderbuffer);
			glRenderbufferStorage(GL_RENDERBUFFER, GL_DEPTH_COMPONENT24, width, height);
			glFramebufferRenderbuffer(GL_FRAMEBUFFER, GL_DEPTH_ATTACHMENT, GL_RENDERBUFFER, depthRenderbuffer);
			
			// Setup matrix. First, flip the y direction because OpenGL textures are not the same way around as CGImages. Then, use ortho to select the part that corresponds to the current tile.
			mat_float16 flipMatrix = (mat_float16) { {1,0,0,0},{0, -1, 0,0}, {0,0,1,0}, {0,0,0,1} };
			mat_float16 combinedMatrix = simd_mul(flipMatrix, cameraMatrix);
			mat_float16 partOfCameraMatrix = simd_orthoMatrix((x/size.width)*2.0-1.0, ((x+width)/size.width)*2.0-1.0, (y/size.height)*2.0-1.0, ((y+height)/size.height)*2.0-1.0, 1, -1);
			combinedMatrix = simd_mul(partOfCameraMatrix, combinedMatrix);
			
			glBindBufferBase(GL_UNIFORM_BUFFER, GLLUniformBlockBindingTransforms, transformBuffer);
			glBufferData(GL_UNIFORM_BUFFER, sizeof(combinedMatrix), NULL, GL_STREAM_DRAW);
            
            mat_float16 *data = glMapBuffer(GL_UNIFORM_BUFFER, GL_WRITE_ONLY);
            memcpy(data, &combinedMatrix, sizeof(combinedMatrix));
            glUnmapBuffer(GL_UNIFORM_BUFFER);
			
			// Enable blend for entire scene. That way, new alpha are correctly combined with values in the buffer (instead of stupidly overwriting them), giving the rendered image a correct alpha channel.
			glEnable(GL_BLEND);
			
			[self drawShowingSelection:NO resetState:YES];
			
			glBindFramebuffer(GL_FRAMEBUFFER, 0);
			
			glFlush();
			
			// Clean up and inform background thread to start loading.
			finishedTextures += 1;
			dispatch_semaphore_signal(texturesReady);
		}
	}
	
	dispatch_semaphore_wait(downloadReady, DISPATCH_TIME_FOREVER);
	glViewport(oldViewport[0], oldViewport[1], oldViewport[2], oldViewport[3]);
	glDeleteFramebuffers(1, &framebuffer);
	glDeleteRenderbuffers(1, &depthRenderbuffer);
	glCullFace(GL_BACK);
	glEnable(GL_MULTISAMPLE);
    
    [[NSNotificationCenter defaultCenter] postNotificationName:GLLDrawStateChangedNotification object:self];
	
	needsUpdateMatrices = YES;
	self.view.needsDisplay = YES;
     */
}

#pragma mark - Private methods

- (void)_updateLights;
{
	struct GLLLightsBuffer lightData;
	
	// Camera position
    lightData.cameraPosition = self.camera.cameraWorldPosition;
	
	// Ambient
	GLLAmbientLight *ambient = (GLLAmbientLight *) lights[0];
	[ambient.color get128BitRGBAComponents:(float*) &lightData.ambientColor];
	
	// Diffuse + Specular
	for (NSUInteger i = 0; i < 3; i++)
	{
		GLLDirectionalLight *light = (GLLDirectionalLight *) lights[i+1];
		lightData.lights[i] = light.uniformBlock;
	}
    
    memcpy(lightBuffer.contents, &lightData, sizeof(lightData));
    [lightBuffer didModifyRange:NSMakeRange(0, sizeof(lightData))];
	
	needsUpdateLights = NO;
}

- (void)_updateMatrices
{
	mat_float16 viewProjection = self.camera.viewProjectionMatrix;
	
	// Set the view projection matrix.
    memcpy(transformBuffer.contents, &viewProjection, sizeof(viewProjection));
    [transformBuffer didModifyRange:NSMakeRange(0, sizeof(viewProjection))];
	
	needsUpdateMatrices = NO;
}

#pragma mark - MTKViewDelegate

- (void)mtkView:(MTKView *)view drawableSizeWillChange:(CGSize)size {
    self.camera.actualWindowWidth = size.width;
    self.camera.actualWindowHeight = size.height;
}

- (void)drawInMTKView:(MTKView *)view {
    MTLRenderPassDescriptor *renderPassDescriptor = view.currentRenderPassDescriptor;
    if (!renderPassDescriptor) {
        return;
    }
    
    id<MTLCommandBuffer> commandBuffer = [commandQueue commandBuffer];
    
    id<MTLRenderCommandEncoder> commandEncoder = [commandBuffer renderCommandEncoderWithDescriptor: renderPassDescriptor];
    
    [commandEncoder setTriangleFillMode:MTLTriangleFillModeFill];
    [commandEncoder setFrontFacingWinding:MTLWindingClockwise];
    [commandEncoder setCullMode:MTLCullModeBack];
    [commandEncoder setDepthStencilState: _sceneDrawer.resourceManager.normalDepthStencilState];
    [commandEncoder setFragmentSamplerState:_sceneDrawer.resourceManager.metalSampler atIndex:0];
    [commandEncoder setBlendColorRed:0.0f green:0.0f blue:0.0f alpha:1.0f];
    
    [self drawShowingSelection: self.view.showSelection into: commandEncoder];
    
    [commandEncoder endEncoding];
    id<MTLDrawable> drawable = view.currentDrawable;
    [commandBuffer presentDrawable:drawable];
    [commandBuffer commit];
}

@end

File: GLLara/GLLViewDrawer.swift
//
//  GLLViewDrawer.swift
//  GLLara
//
//  Created by Torsten Kammer on 06.07.22.
//  Copyright © 2022 Torsten Kammer. All rights reserved.
//

import Foundation
import MetalKit
import UniformTypeIdentifiers

@objc class GLLViewDrawer: NSObject, MTKViewDelegate {
    init(sceneDrawer: GLLSceneDrawer, camera: GLLCamera, view: GLLView) {
        self.sceneDrawer = sceneDrawer
        self.camera = camera
        self.view = view
        
        commandQueue = device.makeCommandQueue()!
        
        // Prepare light buffer.
        lightBuffer = device.makeBuffer(length: MemoryLayout<GLLLightsBuffer>.size, options: .storageModeManaged)!
        lightBuffer.label = "global-lights"
        
        // Load existing lights
        let ambientRequest = NSFetchRequest<GLLAmbientLight>()
        ambientRequest.entity = NSEntityDescription.entity(forEntityName: "GLLAmbientLight", in: sceneDrawer.managedObjectContext!)
        ambientLight = try! sceneDrawer.managedObjectContext!.fetch(ambientRequest)[0];
        
        let directionalLightRequest = NSFetchRequest<GLLDirectionalLight>()
        directionalLightRequest.entity = NSEntityDescription.entity(forEntityName: "GLLDirectionalLight", in: sceneDrawer.managedObjectContext!)
        directionalLightRequest.sortDescriptors = [NSSortDescriptor(key: "index", ascending: true)]
        directionalLights = try! sceneDrawer.managedObjectContext!.fetch(directionalLightRequest)
        
        // Other necessary render state. Thanks to Metal, that got cut down a lot.
        view.clearColor = MTLClearColorMake(0.2, 0.2, 0.2, 1.0)
        
        solidFence = device.makeFence()!
        solidFence.label = "Solid drawing done"
        
        initializeDepthBufferFence = device.makeFence()!
        initializeDepthBufferFence.label = "Depth buffer initialize done"
        
        lastBufferDoneFence = device.makeFence()!
        lastBufferDoneFence.label = "Depth pass done"
        
        let size = view.drawableSize
        camera.actualWindowWidth = Float(size.width)
        camera.actualWindowHeight = Float(size.height)
        surface = Surface(width: Int(size.width), height: Int(size.height), device: device)
        
        super.init()
        
        keyValueObservers.append(camera.observe(\.viewProjectionMatrix) { [weak self] _,_ in
            self?.needsUpdateLights = true
            self?.view?.unpause()
        })
        keyValueObservers.append(ambientLight.observe(\.color) { [weak self] _,_ in
            self?.needsUpdateLights = true
            self?.view?.unpause()
        })
        for light in directionalLights {
            keyValueObservers.append(light.observe(\.uniformBlock) { [weak self] _,_ in
                self?.needsUpdateLights = true
                self?.view?.unpause()
            })
        }
        notificationObservers.append(NotificationCenter.default.addObserver(forName: UserDefaults.didChangeNotification, object: nil, queue: OperationQueue.main) { [weak self] notification in
            self?.updateScaleFactor()
        })
        updateScaleFactor()
        
        view.delegate = self
        
        assert(directionalLights.count == 3, "Only exactly four lights supported at the moment")
    }
    
    let sceneDrawer: GLLSceneDrawer
    let camera: GLLCamera
    weak var view: GLLView?
    
    private let ambientLight: GLLAmbientLight
    private let directionalLights: [GLLDirectionalLight] // Always three, mutations aren't checked
    private let device = GLLResourceManager.shared.metalDevice
    private let commandQueue: MTLCommandQueue
    private var lightBuffer: MTLBuffer
    private var needsUpdateLights = true
    private var keyValueObservers: [NSKeyValueObservation] = []
    private var notificationObservers: [NSObjectProtocol] = []
    
    private var internalBufferScaleFactor = 1.0
    private func updateScaleFactor() {
        let newScaleFactor: Double
        if UserDefaults.standard.bool(forKey: GLLPrefUseMSAA) {
            newScaleFactor = 2.0
        } else {
            newScaleFactor = 1.0
        }
        if newScaleFactor != internalBufferScaleFactor {
            internalBufferScaleFactor = newScaleFactor
            // Recreate textures
            if let view = view {
                let size = view.drawableSize
                surface = Surface(width: Int(size.width * internalBufferScaleFactor), height: Int(size.height * internalBufferScaleFactor), device: device)
            }
        }
    }
    
    // Depth peeling
    private struct Surface {
        var colorTextures: [MTLTexture] = []
        var solidDepthTexture: MTLTexture
        var peelDepthTextures: [MTLTexture] = []
        
        var clearDepthBuffer0PassDescriptor: MTLRenderPassDescriptor
        var solidRenderPassDescriptor: MTLRenderPassDescriptor
        
        let width: Int
        let height: Int
        
        init(width: Int, height: Int, device: MTLDevice) {
            self.width = width
            self.height = height
            
            colorTextures.removeAll()
            let depthPeelLayerCount = 8
            for i in 0..<depthPeelLayerCount {
                let resolvedTextureDescriptor = MTLTextureDescriptor()
                resolvedTextureDescriptor.width = width
                resolvedTextureDescriptor.height = height
                resolvedTextureDescriptor.allowGPUOptimizedContents = true
                resolvedTextureDescriptor.textureType = .type2D
                resolvedTextureDescriptor.pixelFormat = .bgra8Unorm
                resolvedTextureDescriptor.storageMode = .private
                resolvedTextureDescriptor.usage = [ .shaderRead, .renderTarget ]
                
                let texture = device.makeTexture(descriptor: resolvedTextureDescriptor)!
                texture.label = "color-res-\(i)"
                colorTextures.append(texture)
            }
            
            let depthTextureDescriptor = MTLTextureDescriptor()
            depthTextureDescriptor.width = width
            depthTextureDescriptor.height = height
            depthTextureDescriptor.allowGPUOptimizedContents = true
            depthTextureDescriptor.textureType = .type2DMultisample
            depthTextureDescriptor.pixelFormat = .depth32Float
            depthTextureDescriptor.storageMode = .private
            depthTextureDescriptor.textureType = .type2D
            depthTextureDescriptor.usage = [ .shaderRead, .renderTarget ]
            
            solidDepthTexture = device.makeTexture(descriptor: depthTextureDescriptor)!
            solidDepthTexture.label = "depth-solid"
            
            peelDepthTextures.removeAll()
            for i in 0..<2 {
                let depthTexture = device.makeTexture(descriptor: depthTextureDescriptor)!
                depthTexture.label = "depth-\(i)"
                peelDepthTextures.append(depthTexture)
            }
            
            clearDepthBuffer0PassDescriptor = MTLRenderPassDescriptor()
            clearDepthBuffer0PassDescriptor.depthAttachment.texture = peelDepthTextures[0]
            clearDepthBuffer0PassDescriptor.depthAttachment.loadAction = .clear
            clearDepthBuffer0PassDescriptor.depthAttachment.storeAction = .store
            clearDepthBuffer0PassDescriptor.depthAttachment.clearDepth = 0.0
            clearDepthBuffer0PassDescriptor.renderTargetWidth = solidDepthTexture.width
            clearDepthBuffer0PassDescriptor.renderTargetHeight = solidDepthTexture.height

            solidRenderPassDescriptor = MTLRenderPassDescriptor()
            solidRenderPassDescriptor.colorAttachments[0].clearColor = MTLClearColorMake(0.0, 0.0, 0.0, 0.0)
            solidRenderPassDescriptor.colorAttachments[0].texture = colorTextures[0]
            solidRenderPassDescriptor.colorAttachments[0].loadAction = .clear
            solidRenderPassDescriptor.colorAttachments[0].storeAction = .store
            solidRenderPassDescriptor.depthAttachment.texture = solidDepthTexture
            solidRenderPassDescriptor.depthAttachment.loadAction = .clear
            solidRenderPassDescriptor.depthAttachment.storeAction = .store
            solidRenderPassDescriptor.renderTargetWidth = solidDepthTexture.width
            solidRenderPassDescriptor.renderTargetHeight = solidDepthTexture.height
        }
    }
    private var surface: Surface
    
    private var solidFence: MTLFence
    private var initializeDepthBufferFence: MTLFence
    private var lastBufferDoneFence: MTLFence
    
    private var useMultisample = false
    
    private func updateLights() {
        var lightData = GLLLightsBuffer()
        // Camera position
        lightData.cameraPosition = camera.cameraWorldPosition
        
        // Ambient
        lightData.ambientColor = ambientLight.color.rgbaComponents128Bit
        
        // Diffuse + Specular
        lightData.lights.0 = directionalLights[0].uniformBlock
        lightData.lights.1 = directionalLights[1].uniformBlock
        lightData.lights.2 = directionalLights[2].uniformBlock
        
        lightBuffer.contents().copyMemory(from: &lightData, byteCount: MemoryLayout<GLLLightsBuffer>.size)
        lightBuffer.didModifyRange(0 ..< MemoryLayout<GLLLightsBuffer>.size)
        
        needsUpdateLights = false
    }
    
    // MARK: - Image rendering
    // Basic support for render to file
    @objc func writeImage(to url: URL, fileType: UTType, size: CGSize, transparentBackground: Bool) throws {
        // TODO Not yet implemented for metal
        let dataSize = Int(size.width) * Int(size.height) * 4;
        var imageData = Data(count: dataSize);
        imageData.withUnsafeMutableBytes { bytes in
            renderImage(size: size, toColorBuffer: bytes, clearColor: transparentBackground ? MTLClearColorMake(0.0, 0.0, 0.0, 0.0) : MTLClearColorMake(0.2, 0.2, 0.2, 1.0))
        }
        
        let dataProvider = CGDataProvider(data: imageData as CFData)!
        
        let colorSpace = CGColorSpace(name: CGColorSpace.sRGB)!
        let image = CGImage(width: Int(size.width), height: Int(size.height), bitsPerComponent: 8, bitsPerPixel: 32, bytesPerRow: 4 * Int(size.width), space: colorSpace, bitmapInfo: CGBitmapInfo(rawValue: CGImageAlphaInfo.first.rawValue + CGImageByteOrderInfo.order32Little.rawValue), provider: dataProvider, decode: nil, shouldInterpolate: true, intent: .defaultIntent)!
        
        guard let imageDestination = CGImageDestinationCreateWithURL(url as CFURL, fileType.identifier as CFString, 1, nil) else {
            throw NSError(domain: "exporting", code: 1, userInfo: [
                NSLocalizedFailureErrorKey: NSLocalizedString("Could not open file for writing", comment: "Exporting")
            ])
        }
    
        CGImageDestinationAddImage(imageDestination, image, nil)
        guard CGImageDestinationFinalize(imageDestination) else {
            throw NSError(domain: "exporting", code: 1, userInfo: [
                NSLocalizedFailureErrorKey: NSLocalizedString("Could not finalize image file", comment: "Exporting")
            ])
        }
    }
    
    func renderImage(size: CGSize, toColorBuffer colorData: UnsafeMutableRawBufferPointer, clearColor: MTLClearColor) {
        // TODO Not yet implemented in swift and for metal
        
        let surface = Surface(width: Int(size.width), height: Int(size.height), device: GLLResourceManager.shared.metalDevice)
        let queue = device.makeCommandQueue()!
        queue.label = "Write to file queue"
        let commandBuffer = queue.makeCommandBuffer()!
        commandBuffer.label = "Write to file command buffer"
        
        let outputTextureDescriptor = MTLTextureDescriptor()
        outputTextureDescriptor.width = Int(size.width)
        outputTextureDescriptor.height = Int(size.height)
        outputTextureDescriptor.textureType = .type2D
        outputTextureDescriptor.pixelFormat = .bgra8Unorm
        outputTextureDescriptor.usage = [ .renderTarget ]
        // TODO Settings here to improve readback performance. May not be optimal in all cases.
        outputTextureDescriptor.allowGPUOptimizedContents = false
        outputTextureDescriptor.storageMode = .shared
        
        let outputTexture = device.makeTexture(descriptor: outputTextureDescriptor)!
        
        let outputRenderDescriptor = MTLRenderPassDescriptor()
        outputRenderDescriptor.colorAttachments[0].clearColor = clearColor
        outputRenderDescriptor.colorAttachments[0].loadAction = .clear
        outputRenderDescriptor.colorAttachments[0].storeAction = .store
        outputRenderDescriptor.colorAttachments[0].texture = outputTexture
        outputRenderDescriptor.renderTargetWidth = Int(size.width)
        outputRenderDescriptor.renderTargetHeight = Int(size.height)
        
        draw(commandBuffer: commandBuffer, viewRenderPassDescriptor: outputRenderDescriptor, surface: surface, includeUI: false)
        
        commandBuffer.commit()
        commandBuffer.waitUntilCompleted()
        
        outputTexture.getBytes(colorData.baseAddress!, bytesPerRow: Int(size.width) * 4, from: MTLRegionMake2D(0, 0, Int(size.width), Int(size.height)), mipmapLevel: 0)
    }
    
    // MARK: - MTKViewDelegate
    
    func mtkView(_ view: MTKView, drawableSizeWillChange size: CGSize) {
        camera.actualWindowWidth = Float(size.width)
        camera.actualWindowHeight = Float(size.height)
        
        if size.width == 0 || size.height == 0 {
            return;
        }
        
        // Recreate textures
        surface = Surface(width: Int(size.width * internalBufferScaleFactor), height: Int(size.height * internalBufferScaleFactor), device: device)
    }
    
    private func draw(commandBuffer: MTLCommandBuffer, viewRenderPassDescriptor: MTLRenderPassDescriptor, surface: Surface, includeUI: Bool = true) {
        
        var viewProjection = camera.viewProjectionMatrix(forAspectRatio: Float(surface.width) / Float(surface.height))
        
        if needsUpdateLights {
            updateLights()
        }
        
        // Step 1: Render everything solid, with depth buffer 0 as normal depth buffer, to texture 0.
        let solidPassEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: surface.solidRenderPassDescriptor)!
        solidPassEncoder.label = "Draw solids"
        solidPassEncoder.setCullMode(.back)
        solidPassEncoder.setDepthStencilState(sceneDrawer.resourceManager.normalDepthStencilState)
        solidPassEncoder.setFragmentSamplerState(sceneDrawer.resourceManager.metalSampler, index: 0)
        solidPassEncoder.setVertexBytes(&viewProjection, length: MemoryLayout<float4x4>.size, index: Int(GLLVertexInputIndexViewProjection.rawValue))
        solidPassEncoder.setVertexBuffer(lightBuffer, offset: 0, index: Int(GLLVertexInputIndexLights.rawValue))
        solidPassEncoder.setFragmentBuffer(lightBuffer, offset: 0, index: Int(GLLFragmentBufferIndexLights.rawValue))
        
        sceneDrawer.draw(into: solidPassEncoder, blended: false)
        
        solidPassEncoder.updateFence(solidFence, after: [.fragment])
        solidPassEncoder.endEncoding()


        // Step 2: For every further resolved texture we have:
        // - Use previous depth buffer as peel front buffer (only things behind it get drawn)
        // - Use other depth buffer as normal depth buffer, but initialized to depth buffer from solid
        // - Draw alpha, into multisample texture, resolving to resolved texture i
        
        let clearDepthBuffer0Pass = commandBuffer.makeRenderCommandEncoder(descriptor: surface.clearDepthBuffer0PassDescriptor)!
        clearDepthBuffer0Pass.label = "Clear Depth Buffer 0"
        clearDepthBuffer0Pass.updateFence(lastBufferDoneFence, after: [.fragment])
        clearDepthBuffer0Pass.endEncoding()
        
        let depthPeelPassDescriptor = MTLRenderPassDescriptor()
        depthPeelPassDescriptor.colorAttachments[0].clearColor = MTLClearColorMake(0.0, 0.0, 0.0, 0.0)
        depthPeelPassDescriptor.colorAttachments[0].loadAction = .clear
        depthPeelPassDescriptor.colorAttachments[0].storeAction = .store
        depthPeelPassDescriptor.depthAttachment.loadAction = .load
        depthPeelPassDescriptor.depthAttachment.storeAction = .store
        depthPeelPassDescriptor.renderTargetWidth = surface.solidDepthTexture.width
        depthPeelPassDescriptor.renderTargetHeight = surface.solidDepthTexture.height

        
        var lastWrittenDepthBuffer = 0
        for i in 1 ..< surface.colorTextures.count {
            let backDepthBuffer = 1 - lastWrittenDepthBuffer
            let isLast = i + 1 == surface.colorTextures.count
            
            let initializeDepthBufferEncoder = commandBuffer.makeBlitCommandEncoder()!
            initializeDepthBufferEncoder.label = "Initialize Depth Buffer \(i)"
            initializeDepthBufferEncoder.waitForFence(lastBufferDoneFence)
            if i == 1 {
                initializeDepthBufferEncoder.waitForFence(solidFence)
            }
            initializeDepthBufferEncoder.copy(from: surface.solidDepthTexture, to: surface.peelDepthTextures[backDepthBuffer])
            initializeDepthBufferEncoder.updateFence(initializeDepthBufferFence)
            initializeDepthBufferEncoder.endEncoding()
            
            depthPeelPassDescriptor.colorAttachments[0].texture = surface.colorTextures[i]
            depthPeelPassDescriptor.depthAttachment.texture = surface.peelDepthTextures[backDepthBuffer]
            if isLast {
                depthPeelPassDescriptor.depthAttachment.storeAction = .dontCare
            }

            let depthPeelPassEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: depthPeelPassDescriptor)!
            depthPeelPassEncoder.label = "Depth Peel Layer \(i)"
            depthPeelPassEncoder.waitForFence(initializeDepthBufferFence, before: [ .vertex ])
            depthPeelPassEncoder.setCullMode(.back)
            depthPeelPassEncoder.setDepthStencilState(sceneDrawer.resourceManager.normalDepthStencilState)
            depthPeelPassEncoder.setFragmentSamplerState(sceneDrawer.resourceManager.metalSampler, index: 0)
            depthPeelPassEncoder.setVertexBytes(&viewProjection, length: MemoryLayout<float4x4>.size, index: Int(GLLVertexInputIndexViewProjection.rawValue))
            depthPeelPassEncoder.setVertexBuffer(lightBuffer, offset: 0, index: Int(GLLVertexInputIndexLights.rawValue))
            depthPeelPassEncoder.setFragmentBuffer(lightBuffer, offset: 0, index: Int(GLLFragmentBufferIndexLights.rawValue))
            
            depthPeelPassEncoder.setFragmentTexture(surface.peelDepthTextures[lastWrittenDepthBuffer], index: Int(GLLFragmentArgumentIndexTextureDepthPeelFront.rawValue))
            
            sceneDrawer.draw(into: depthPeelPassEncoder, blended: true)
            
            depthPeelPassEncoder.updateFence(lastBufferDoneFence, after: [ .fragment ])
            depthPeelPassEncoder.endEncoding()
            
            lastWrittenDepthBuffer = backDepthBuffer
        }
        
        // Step 3: Using the view render pass descriptor, render all resolved textures on top of each other with blending.
        let combineCommandEncoder = commandBuffer.makeRenderCommandEncoder(descriptor: viewRenderPassDescriptor)!
        
        combineCommandEncoder.label = "Final combine"
        combineCommandEncoder.waitForFence(lastBufferDoneFence, before: [.vertex])
        combineCommandEncoder.setRenderPipelineState(sceneDrawer.resourceManager.squarePipelineState)
        combineCommandEncoder.setVertexBuffer(sceneDrawer.resourceManager.squareVertexArray, offset: 0, index: 0)
        
        // Order: 0, n-1, n-2, ..., 1
        combineCommandEncoder.setFragmentTexture(surface.colorTextures[0], index: 0)
        combineCommandEncoder.drawPrimitives(type: .triangleStrip, vertexStart: 0, vertexCount: 4)
        for i in 1 ..< surface.colorTextures.count {
            combineCommandEncoder.setFragmentTexture(surface.colorTextures[surface.colorTextures.count - i], index: 0)
            combineCommandEncoder.drawPrimitives(type: .triangleStrip, vertexStart: 0, vertexCount: 4)
        }
        
        // Step 3.5: If present, draw the skeleton view on top of all that.
        if includeUI {
            if self.view?.showSelection ?? false {
                combineCommandEncoder.setVertexBytes(&viewProjection, length: MemoryLayout<float4x4>.size, index: Int(GLLVertexInputIndexViewProjection.rawValue))
                combineCommandEncoder.setVertexBuffer(lightBuffer, offset: 0, index: Int(GLLVertexInputIndexLights.rawValue))
                combineCommandEncoder.setFragmentBuffer(lightBuffer, offset: 0, index: Int(GLLFragmentBufferIndexLights.rawValue))
                
                sceneDrawer.drawSelection(int: combineCommandEncoder)
            }
        }
        
        combineCommandEncoder.endEncoding()
    }
    
    func draw(in view: MTKView) {
        guard let viewRenderPassDescriptor = view.currentRenderPassDescriptor else {
            return
        }
        
        guard let commandBuffer = commandQueue.makeCommandBuffer() else {
            return
        }
        
        draw(commandBuffer: commandBuffer, viewRenderPassDescriptor: viewRenderPassDescriptor, surface: surface, includeUI: true)
        
        let drawable = view.currentDrawable
        commandBuffer.present(drawable!)
        commandBuffer.commit()
    }
}
